### PR TITLE
feat: replace top-level this in esm

### LINKED ISSUE #4421: should replace top-level this to undefined in esm
https://github.com/webpack/webpack/blob/87660921808566ef3b8796f8df61bd79fc026108/lib/dependencies/HarmonyTopLevelThisParserPlugin.js#L22

- webpack-test/cases/parsing/es2020

### PR DESCRIPTION
**Pull request by Austaras**
## Summary

Closes #4421

## Require Documentation?

- [x] No
- [ ] Yes, the corresponding rspack-website PR is \_\_

### CODE REVIEW

**Comment by ahabhgk** [comment]
I think now `webpack-test/cases/parsing/es2020` can pass, could you enable it, just need to make `test.filter.js` return true

**Comment by IWANABETHATGUY** [comment]
> I think now `webpack-test/cases/parsing/es2020` can pass, could you enable it, just need to make `test.filter.js` return true, and run `pnpm run test -- -t "parsing es2020"` under `webpack-test` folder

string replacing-based codegen required ast to be untouched (some plugins needs span info), if we add an extra transforming pass, it may affect the optimization result. Can we move the transforming work to the loader side?

**Comment by IWANABETHATGUY** [comment]
Or just like the webpack, use a`ConstDependency` instead.

**Comment by Austaras** [comment]
Turns out the origin impl doesn't work at all. Two separate AST is used for codegen and dep analysis, and modify the latter one doesn't have any effect.

**Comment by Austaras** [comment]
> > I think now `webpack-test/cases/parsing/es2020` can pass, could you enable it, just need to make `test.filter.js` return true, and run `pnpm run test -- -t "parsing es2020"` under `webpack-test` folder
> 
> string replacing-based codegen required ast to be untouched (some plugins needs span info), if we add an extra transforming pass, it may affect the optimization result. Can we move the transforming work to the loader side?

Seems like most passes in `transform_before_pass` belong in the loader, but for this PR I believe it's better not to change this logic.

File: crates/rspack_plugin_javascript/src/visitors/mod.rs [hunk 1]
```diff
@@ -240,6 +242,7 @@ pub fn run_before_pass(
       builtins_webpack_plugin(options, unresolved_mark),
       // This will be deprecated in the future when TypeScript support and transform by default is dropped.
       // These features are coupled with this.
+      replace_top_level_this(unresolved_mark),
```
Comment (ahabhgk) [inline crates/rspack_plugin_javascript/src/visitors/mod.rs:245]: And move this under this branch https://github.com/web-infra-dev/rspack/blob/3c0c6ca9d9cc5f99919520288006e33f14867e57/crates/rspack_plugin_javascript/src/visitors/dependency/mod.rs#L131

**Review by ahabhgk — COMMENTED** [review]
It would be better if you can implement it by string-replacement way (use `ConstDependency` to replace `this` to `undefined`), this is pretty similar with `ReplaceNestWebpackRequireVisitor`, so you can reference it at https://github.com/web-infra-dev/rspack/blob/3c0c6ca9d9cc5f99919520288006e33f14867e57/crates/rspack_plugin_javascript/src/visitors/dependency/compatibility_scanner.rs#L52-L75

**Comment by Austaras** [comment]
Done. But what's the difference between AST plugin and text plugin? What's the necessity of the latter?

**Comment by ahabhgk** [comment]
We used ast-based replacement for all our previous implementations, but we found it challenging to implement caching and future persistent caching using ast compared to string replacement. Therefore, we switched to the implementation of string replacement, which is currently in an intermediate stage. We currently have both string replacement and ast replacement, as there are still some blockers for the remaining ast replacement. In the long term, we will replace all of it with string replacement.

**Comment by ahabhgk** [comment]
LGTM, can you make the CI passes

**Comment by Austaras** [comment]
Interesting.

**Review by ahabhgk — APPROVED** [review]

**Comment by ahabhgk** [comment]
Thanks!

**Pull request merged**